lib/engine/game/g_1848/step/buy_train.rb

```diff
@@ -50,25 +50,31 @@ def owns_2e?(entity)
           end
 
           def can_buy_2e?(entity)
-            ghan_private_owned?(entity) &&
-            @game.phase.available?(ghan_train.available_on) &&
-            ghan_train.price - ghan_private_ability.discount <= entity.cash
+            return false if !@game.phase.available?(ghan_train&.available_on) || owns_2e?(entity)
+
+            cost = ghan_train.price
+            cost -= ghan_private_ability.discount if ghan_private_owned?(entity)
+            cost <= entity.cash
           end
 
           def ghan_private_ability
             @ghan_private_ability ||= @game.abilities(@game.ghan, :train_discount, time: ability_timing)
           end
 
           def ghan_train
-            @ghan_train ||= @depot.trains.find { |t| t.name == '2E' }
+            @depot.depot_trains.find { |t| t.name == '2E' }
           end
 
           def ghan_private_owned?(entity)
             entity.companies.include?(@game.ghan) || entity.owner.companies.include?(@game.ghan)
           end
 
+          def at_train_limit?(entity)
+            entity.trains.count { |t| t.name != '2E' } == @game.train_limit(entity)
+          end
+
           def room?(entity)
-            return true if entity.trains.count { |t| t.name != '2E' } == @game.train_limit(entity) && can_buy_2e?(entity)
+            return true if at_train_limit?(entity) && can_buy_2e?(entity)
 
             entity.trains.count { |t| t.name != '2E' } < @game.train_limit(entity)
```

Comment:
New methods added simplify this

def room?(entity) 
  !at_train_limit?(entity) || can_buy_2e?(entity)
end